File: pkg-py/tests/test_remote_rfam.py
"""Rfam remote mode: live existence checks via the Rfam API."""

import pytest

import biogate
import biogate._remote as remote
from biogate import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_builds_the_rfam_family_endpoint():
    from biogate._registry import get_source

    url = remote._rfam_url(get_source("rfam"), "RF00001")
    assert url == "https://rfam.org/family/RF00001?content-type=application/json"


def test_existing_family_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("RF00001", source_db="rfam", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "RF00001"
    assert res.suggestion is None


def test_absent_family_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biogate.check_id("RF99999", source_db="rfam", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("rf00001", source_db="rfam", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "RF00001"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biogate.check_id("RF001", source_db="rfam", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biogate.check_id("RF00001", source_db="rfam", how="remote")
    assert not remote._remote_cache_path("rfam", "family", "RF00001").is_file()

File: pkg-r/tests/testthat/test-remote-rfam.R
# Offline tests for rfam remote mode. Existence is checked against the Rfam API;
# the biobouncer.remote_transport option replaces it.

.stub_rfam <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url addresses the rfam family endpoint", {
  url <- .remote_resolvers$rfam$url(NULL, "RF00001")
  expect_identical(
    url,
    "https://rfam.org/family/RF00001?content-type=application/json"
  )
})

test_that("an existing family is valid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_rfam(200))
  res <- check_id("RF00001", "rfam", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "RF00001")
  expect_true(is.na(res$suggestion))
})

test_that("an absent family is invalid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_rfam(404))
  res <- check_id("RF99999", "rfam", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_rfam(200))
  res <- check_id("rf00001", "rfam", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "RF00001")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("RF001", "rfam", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_rfam(500))
  expect_error(
    check_id("RF00001", "rfam", how = "remote"),
    class = "biobouncer_error_remote"
  )
  expect_false(file.exists(.remote_cache_path("rfam", "family", "RF00001")))
})
